### DISCUSSION TITLE
Background agend missing when installing from source

### DISCUSSION THREAD

**Discussion by gforien** [pqrs-org/Karabiner-Elements#4022] [Q&A] [answered]
Hi everyone
I'm trying to install Karabiner-Elements from source.

After a successful build (unsigned), I double-click on the .dmg file and follow the installation process.

The background agents are missing so I ran
```sh
sudo launchctl load /Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Privileged Daemons.app/Contents/Library/LaunchDaemons/*.plist
sudo launchctl load /Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Daemons.app/Contents/Library/LaunchDaemons/*.plist

# -> Warning: Expecting a LaunchDaemons path since the command was ran as root. Got LaunchAgents instead.
# -> launchctl bootstrap is a recommended alternative
```

I get a warning, but this should not be a problem I suppose.
Yet, the background agents are still missing.

What am I missing ?

Related documentation
- https://karabiner-elements.pqrs.org/docs/help/advanced-topics/installed-files/

thanks for your help

**Comment by tekezo** [comment, marked as answer]
These daemons and agents are designed to be loaded by apps rather than launchctl.
Enable them by the following command. (/Applications/Karabiner-Elements.app executes the same command internally) 

```shell
'/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Privileged Daemons.app/Contents/MacOS/Karabiner-Elements Privileged Daemons' register-core-daemons

'/Library/Application Support/org.pqrs/Karabiner-Elements/Karabiner-Elements Non-Privileged Agents.app/Contents/MacOS/Karabiner-Elements Non-Privileged Agents' register-core-agents
```

**Comment by tekezo** [comment]
Furthermore, any permissions granted in macOS System Settings will become invalid if a binary signed by a different signer is installed.
(For unsigned binaries, permissions will be invalidated each time the binary is updated.)

In System Settings, it may look like the permissions are properly granted even when they have been invalidated. I recommend removing all permissions once and redoing the process starting from the permission dialog. You can find the list of required permissions on this page:
https://karabiner-elements.pqrs.org/docs/manual/misc/required-macos-settings/

**Comment by gforien** [comment]
Wow I did not expect an answer from yourself!
Sorry for the delayed response. Just wanted to say thank you because I've been using Karabiner for 3 years now and I could not go back to using macOS without it. I have 600+ rules across 3 profiles. It's truly useful in all kinds of situations
